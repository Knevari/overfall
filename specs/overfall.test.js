const Overfall = require('../src');

it('should be able to update state', function() {
  const overfall = new Overfall();

  overfall.changeState({ foo: "bar" });
  overfall.changeState({ bar: "zaz" });

  expect(overfall.state.foo).toBe('bar');
})

it('should be able to create events', function() {
  const overfall = new Overfall();

  overfall.on("event_one").do(() => {
    console.log("First event published");
  })

  overfall.on("event_two").do(() => {
    console.log("Second event published");
  })

  overfall.on("event_three").do(() => {
    console.log("Third event published");
  })

  expect(overfall.eventBus.events).toMatchSnapshot();
})

it('should remove the event dependency if the dependant ' +
   'piece of state is removed', function() {
  const overfall = new Overfall({ books: [], movies: [] });

  overfall.on("event").do(() => {}).when("books", "movies");
  overfall.changeState(previous => ({ movies: [] }));

  expect(overfall.eventBus.events.event.dependencies).toContain("movies");
})

it('should execute the event functions correctly if' +
   'some dependency is removed', function() {
   const overfall = new Overfall({ books: [], movies: [] });

   let executedFirst = false;
   let executedSecond = false;

   overfall
    .on("event_one")
    .do(() => (executedFirst = true))
    .when("books", "movies");
   overfall.changeState(previous => ({ movies: [] }));

   overfall
    .on("event_two")
    .do(() => (executedSecond = true))
    .when("movies");
   overfall.changeState({ movies: ["Lord of The Rings"] });

   expect(executedFirst).toBe(true);
   expect(executedSecond).toBe(true);
})
