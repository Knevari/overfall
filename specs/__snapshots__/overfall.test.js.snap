// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be able to create events 1`] = `
Object {
  "event_one": Object {
    "dependencies": Array [],
    "subscribers": Object {
      Symbol(id): [Function],
    },
  },
  "event_three": Object {
    "dependencies": Array [],
    "subscribers": Object {
      Symbol(id): [Function],
    },
  },
  "event_two": Object {
    "dependencies": Array [],
    "subscribers": Object {
      Symbol(id): [Function],
    },
  },
}
`;
